File: packages-gen/js/event/base-form.js
const event_common = require("@xtalk/event/base-listener.js")

const xtd = require("@xtalk/lang/common-data.js")

const validate = require("@xtalk/event/util-validate.js")

function make_form(initial,validators){
  let result = validate.create_result(validators);
  return event_common.make_container(
    initial,
    "event.form",
    {"result":result,"validators":validators}
  );
}

function check_event(event,fields){
  for(let field of fields){
    for(let evfield of event["fields"]){
      if(evfield == field){
        return true;
      }
    };
  };
  return false;
}

function add_listener(form,listener_id,fields,callback,meta){
  fields = event_common.arrayify_path(fields);
  return event_common.add_listener(form,listener_id,"form",callback,Object.assign({"form/fields":fields},meta),function (event){
    return check_event(event,fields);
  });
}

var remove_listener = event_common.remove_listener;

var list_listeners = event_common.list_listeners;

function trigger_all(form,event_type){
  let {validators} = form;
  let fields = Object.keys(validators);
  return event_common.trigger_listeners(form,{"type":event_type,"fields":fields});
}

function trigger_field(form,fields,event_type){
  return event_common.trigger_listeners(form,{
    "type":event_type,
    "fields":event_common.arrayify_path(fields)
  });
}

function set_field(form,field,value){
  let {data} = form;
  data[field] = value;
  return trigger_field(form,field,"form.data");
}

function get_field(form,field){
  let {data} = form;
  return data[field];
}

function toggle_field(form,field){
  return set_field(form,field,!get_field(form,field));
}

function field_fn(form,field){
  return function (value){
    return set_field(form,field,value);
  };
}

function get_result(form){
  return form["result"];
}

function get_field_result(form,field){
  let {result} = form;
  let {fields} = result;
  return fields[field];
}

function get_data(form){
  return form["data"];
}

function set_data(form,m){
  let {data} = form;
  Object.assign(data,m);
  let fields = Object.keys(m);
  return trigger_field(form,fields,"form.data");
}

function reset_all_data(form){
  let {initial} = form;
  let data = initial();
  form["data"] = data;
  return trigger_all(form,"form.data");
}

function reset_field_data(form,field){
  let {data,initial} = form;
  let value = (initial())[field];
  data[field] = value;
  return trigger_field(form,field,"form.data");
}

function validate_all(form,hook_fn,complete_fn){
  let {data,result,validators} = form;
  return validate.validate_all(data,validators,result,hook_fn,null).then(function (res){
    trigger_all(form,"form.validation");
    if(complete_fn){
      complete_fn("ok" == res["status"],res);
    }
    return res;
  });
}

function validate_field(form,field,hook_fn,complete_fn){
  let {data,result,validators} = form;
  return validate.validate_field(data,field,validators,result,hook_fn,null).then(function (res){
    trigger_field(form,field,"form.validation");
    if(complete_fn){
      complete_fn("ok" == xtd.get_in(res["fields"],[field,"status"]),res);
    }
    return res;
  });
}

function reset_field_validator(form,field){
  let {result} = form;
  result[field] = {"status":"pending"};
  trigger_field(form,field,"form.validation");
  return result;
}

function reset_all_validators(form){
  let {result,validators} = form;
  form["result"] = validate.create_result(validators);
  trigger_all(form,"form.validation");
  return result;
}

function reset_all(form){
  reset_all_data(form);
  reset_all_validators(form);
}

function check_field_passed(form,field){
  let {result} = form;
  let {fields} = result;
  return "ok" == xtd.get_in(fields,[field,"status"]);
}

function check_field_errored(form,field){
  let {result} = form;
  let {fields} = result;
  return "errored" == xtd.get_in(fields,[field,"status"]);
}

function check_all_passed(form){
  let {result} = form;
  let {fields} = result;
  for(let v of Object.values(fields)){
    if("ok" != v["status"]){
      return false;
    }
  };
  return true;
}

function check_any_errored(form){
  let {result} = form;
  let {fields} = result;
  for(let v of Object.values(fields)){
    if("errored" == v["status"]){
      return true;
    }
  };
  return false;
}

module.exports = {
  ["make_form"]:make_form,
  ["check_event"]:check_event,
  ["add_listener"]:add_listener,
  ["remove_listener"]:remove_listener,
  ["list_listeners"]:list_listeners,
  ["trigger_all"]:trigger_all,
  ["trigger_field"]:trigger_field,
  ["set_field"]:set_field,
  ["get_field"]:get_field,
  ["toggle_field"]:toggle_field,
  ["field_fn"]:field_fn,
  ["get_result"]:get_result,
  ["get_field_result"]:get_field_result,
  ["get_data"]:get_data,
  ["set_data"]:set_data,
  ["reset_all_data"]:reset_all_data,
  ["reset_field_data"]:reset_field_data,
  ["validate_all"]:validate_all,
  ["validate_field"]:validate_field,
  ["reset_field_validator"]:reset_field_validator,
  ["reset_all_validators"]:reset_all_validators,
  ["reset_all"]:reset_all,
  ["check_field_passed"]:check_field_passed,
  ["check_field_errored"]:check_field_errored,
  ["check_all_passed"]:check_all_passed,
  ["check_any_errored"]:check_any_errored
}

File: packages-gen/dart/libs/xt-event/lib/base-form.dart
import 'package:xtalk_event/base-listener.dart' as event_common;
import 'package:xtalk_lang/common-data.dart' as xtd;
import 'package:xtalk_event/util-validate.dart' as validate;
import 'dart:async';




make_form(initial, validators) {
  var result = validate.create_result(validators);
  return event_common.make_container(
    initial,
    "event.form",
    <dynamic, dynamic>{"result":result,"validators":validators}
  );
}

check_event(event, fields) {
  var arr_51011 = fields;
  for(var i51012 = 0; i51012 < arr_51011.length; ++i51012){
    var field = arr_51011[i51012];
    var arr_51033 = event["fields"];
    for(var i51034 = 0; i51034 < arr_51033.length; ++i51034){
      var evfield = arr_51033[i51034];
      if(evfield == field){
        return true;
      }
    };
  };
  return false;
}

add_listener(form, listener_id, fields, callback, meta) {
  fields = event_common.arrayify_path(fields);
  return event_common.add_listener(form,listener_id,"form",callback,xtd.obj_assign(<dynamic, dynamic>{"form/fields":fields},meta),(event) {
    return check_event(event,fields);
  });
}

var remove_listener = event_common.remove_listener;

var list_listeners = event_common.list_listeners;

trigger_all(form, event_type) {
  var validators = form["validators"];
  var fields = List<dynamic>.from(( validators ).keys);
  return event_common.trigger_listeners(form,<dynamic, dynamic>{"type":event_type,"fields":fields});
}

trigger_field(form, fields, event_type) {
  return event_common.trigger_listeners(form,<dynamic, dynamic>{
    "type":event_type,
    "fields":event_common.arrayify_path(fields)
  });
}

set_field(form, field, value) {
  var data = form["data"];
  data[field] = value;
  return trigger_field(form,field,"form.data");
}

get_field(form, field) {
  var data = form["data"];
  return data[field];
}

toggle_field(form, field) {
  return set_field(form,field,!get_field(form,field));
}

field_fn(form, field) {
  return (value) {
    return set_field(form,field,value);
  };
}

get_result(form) {
  return form["result"];
}

get_field_result(form, field) {
  var result = form["result"];
  var fields = result["fields"];
  return fields[field];
}

get_data(form) {
  return form["data"];
}

set_data(form, m) {
  var data = form["data"];
  xtd.obj_assign(data,m);
  var fields = List<dynamic>.from(( m ).keys);
  return trigger_field(form,fields,"form.data");
}

reset_all_data(form) {
  var initial = form["initial"];
  var data = initial();
  form["data"] = data;
  return trigger_all(form,"form.data");
}

reset_field_data(form, field) {
  var data = form["data"];
  var initial = form["initial"];
  var value = (initial())[field];
  data[field] = value;
  return trigger_field(form,field,"form.data");
}

validate_all(form, hook_fn, complete_fn) {
  var data = form["data"];
  var result = form["result"];
  var validators = form["validators"];
  return ((Future.sync(() => validate.validate_all(data,validators,result,hook_fn,null))) as Future<dynamic>).then((value) async { return await Function.apply((res) {
    trigger_all(form,"form.validation");
    if((null != complete_fn) && (false != complete_fn)){
      Function.apply(
        (complete_fn as Function),
        <dynamic>["ok" == res["status"],res]
      );
    }
    return res;
  },<dynamic>[value]); });
}

validate_field(form, field, hook_fn, complete_fn) {
  var data = form["data"];
  var result = form["result"];
  var validators = form["validators"];
  return ((Future.sync(() => validate.validate_field(data,field,validators,result,hook_fn,null))) as Future<dynamic>).then((value) async { return await Function.apply((res) {
    trigger_field(form,field,"form.validation");
    if((null != complete_fn) && (false != complete_fn)){
      Function.apply((complete_fn as Function),<dynamic>[
        "ok" == xtd.get_in(res["fields"],<dynamic>[field,"status"]),
        res
      ]);
    }
    return res;
  },<dynamic>[value]); });
}

reset_field_validator(form, field) {
  var result = form["result"];
  result[field] = <dynamic, dynamic>{"status":"pending"};
  trigger_field(form,field,"form.validation");
  return result;
}

reset_all_validators(form) {
  var result = form["result"];
  var validators = form["validators"];
  form["result"] = validate.create_result(validators);
  trigger_all(form,"form.validation");
  return result;
}

reset_all(form) {
  reset_all_data(form);
  return reset_all_validators(form);
}

check_field_passed(form, field) {
  var result = form["result"];
  var fields = result["fields"];
  return "ok" == xtd.get_in(fields,<dynamic>[field,"status"]);
}

check_field_errored(form, field) {
  var result = form["result"];
  var fields = result["fields"];
  return "errored" == xtd.get_in(fields,<dynamic>[field,"status"]);
}

check_all_passed(form) {
  var result = form["result"];
  var fields = result["fields"];
  for(var v in fields.values){
    if("ok" != v["status"]){
      return false;
    }
  };
  return true;
}

check_any_errored(form) {
  var result = form["result"];
  var fields = result["fields"];
  for(var v in fields.values){
    if("errored" == v["status"]){
      return true;
    }
  };
  return false;
}